internal/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection_test.go

```diff
@@ -0,0 +1,191 @@
+			Name:      leaseName,
+			Namespace: namespace,
+		},
+	}
+	return lease
+}
+
+// applyLease applies the lease
+func applyLease(client kubernetes.Interface, lease *v1.Lease, firstTime bool) error {
+	var err error
+	if firstTime {
+		_, err = client.CoordinationV1().Leases(namespace).Create(context.Background(), lease, metav1.CreateOptions{})
+		return err
+	}
+	_, err = client.CoordinationV1().Leases(namespace).Update(context.Background(), lease, metav1.UpdateOptions{})
+	return err
+}
+
+// getLeaseHolder returns the holder identity of the lease
+func getLeaseHolder(client kubernetes.Interface) (string, error) {
+	lease, err := client.CoordinationV1().Leases(namespace).Get(context.Background(), leaseName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	holder := lease.Spec.HolderIdentity
+	if holder == nil {
+		return "", err
+	}
+	return *holder, nil
+}
+
+// TestNewLeaderElectionManager will test the leader elector.
+// We will try to check if an instance can acquire the lease more than one time. This way, we will know that
+// the leader elector starts running again after it has stopped - which happens once a leader looses the lease.
+// To make sure that happens we will do the following:
+// 1. We will create the lease to be used by the leader elector.
+// 2. We will create two context providers - in the default context, this would mean two nodes, each one with an agent running.
+// We will wait for one of the agents, agent1, to acquire the lease, before starting the other.
+// 3. We force the lease to be acquired by the other agent, agent2.
+// 4. We will force the lease to be acquired by the agent1 again. To avoid the agent2 reacquiring it multiple times,
+// we will stop this provider and make sure the agent1 can reacquire it.
+func TestNewLeaderElectionManager(t *testing.T) {
+	client := k8sfake.NewSimpleClientset()
+
+	lease := createLease()
+	// create the lease that leader election will be using
+	err := applyLease(client, lease, true)
+	require.NoError(t, err)
+
+	// Create the provider
+	logger := logp.NewLogger("test_leaderelection")
+
+	leaseDuration := 3
+	leaseRenewDeadline := 2
+	leaseRetryPeriod := 1
+
+	c := map[string]interface{}{
+		"leader_lease":         leaseName,
+		"leader_leaseduration": leaseDuration,
+		"leader_renewdeadline": leaseRenewDeadline,
+		"leader_retryperiod":   leaseRetryPeriod,
+	}
+	cfg, err := config.NewConfigFrom(c)
+	require.NoError(t, err)
+
+	getK8sClientFunc = func(kubeconfig string, opt autodiscoverK8s.KubeClientOptions) (kubernetes.Interface, error) {
+		return client, nil
+	}
+	require.NoError(t, err)
+
+	podNames := [2]string{"agent1", "agent2"}
+	cancelFuncs := [2]context.CancelFunc{}
+
+	// Create two leader election providers representing two agents running
+	for i := 0; i < 2; i++ {
+		p, err := ContextProviderBuilder(logger, cfg, true)
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		cancelFuncs[i] = cancel
```

Comment:
You should probably still have `defer cancel()`.

If a below `require.*` fails the test will end immediately and those contexts will not be cancelled. I know its just tests, but being pragmatic is always good.